src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/ApplyVQSR.java

```diff
@@ -0,0 +1,628 @@
+    private boolean lenientVCFProcessing;
+
+    /**
+     * For this to work properly, the -ignoreFilter argument should also be applied to the VariantRecalibration command.
+     */
+    @Argument(fullName="ignore_filter", shortName="ignoreFilter", doc="If specified, the recalibration will be applied to variants marked as filtered by the specified filter name in the input VCF file", optional=true)
+    private List<String> IGNORE_INPUT_FILTERS = new ArrayList<>();
+
+    @Argument(fullName="ignore_all_filters", shortName="ignoreAllFilters", doc="If specified, the variant recalibrator will ignore all input filters. Useful to rerun the VQSR from a filtered output file.", optional=true)
+    private boolean IGNORE_ALL_FILTERS = false;
+
+    @Argument(fullName="excludeFiltered", shortName="ef", doc="Don't output filtered loci after applying the recalibration", optional=true)
+    private boolean EXCLUDE_FILTERED = false;
+
+    @Argument(fullName = "mode", shortName = "mode", doc = "Recalibration mode to employ: 1.) SNP for recalibrating only SNPs (emitting indels untouched in the output VCF); 2.) INDEL for indels; and 3.) BOTH for recalibrating both SNPs and indels simultaneously.", optional=true)
+    private VariantRecalibratorArgumentCollection.Mode MODE = VariantRecalibratorArgumentCollection.Mode.SNP;
+
+    /////////////////////////////
+    // Private Member Variables
+    /////////////////////////////
+    private VariantContextWriter vcfWriter;
+    final private List<Tranche> tranches = new ArrayList<>();
+    final private Set<String> inputNames = new HashSet<>();
+    final private Set<String> ignoreInputFilterSet = new TreeSet<>();
+    final static private String listPrintSeparator = ",";
+    final static private String trancheFilterString = "VQSRTranche";
+    final static private String arrayParseRegex = "[\\[\\]\\s]";
+    final static private String emptyStringValue = "NA";
+    final static private String emptyFloatValue = "NaN";
+
+
+    //---------------------------------------------------------------------------------------------------------------
+    //
+    // initialize
+    //
+    //---------------------------------------------------------------------------------------------------------------
+
+    public void onTraversalStart() {
+        if( TS_FILTER_LEVEL != null ) {
+            try {
+                for (final Tranche t : Tranche.readTranches(TRANCHES_FILE)) {
+                    if (t.targetTruthSensitivity >= TS_FILTER_LEVEL) {
+                        tranches.add(t);
+                    }
+                    logger.info(String.format("Read tranche " + t));
+                }
+            }
+            catch(IOException e ) {
+                throw new UserException.CouldNotReadInputFile(TRANCHES_FILE, e);
+            }
+            Collections.reverse(tranches); // this algorithm wants the tranches ordered from best (lowest truth sensitivity) to worst (highest truth sensitivity)
+        }
+
```

Comment:
Is `inputNames` not getting initialized? In GATK3 it was initialized with a loop like this:

```
        for( final RodBinding rod : input ) {
            inputNames.add( rod.getName() );
        }
```